web_widget_domain_editor_dialog/static/src/js/widget_domain_editor_dialog.js

```diff
@@ -0,0 +1,74 @@
+/* Copyright 2019 Tecnativa - David Vidal
+ * License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). */
+
+odoo.define("web_widget_domain_editor_dialog.DomainEditorDialog", function (require) {
+    "use strict";
+
+    const core = require("web.core");
+    const view_dialogs = require("web.view_dialogs");
+    const Domain = require("web.Domain");
+    const _t = core._t;
+
+    const DomainEditorDialog = view_dialogs.SelectCreateDialog.extend({
+        init: function () {
+            this._super.apply(this, arguments);
+            const _this = this;
+            this.options = _.defaults(this.options, {
+                dynamicFilters: [
+                    {
+                        description: _.str.sprintf(_t("Selected domain")),
+                        domain: Domain.prototype.stringToArray(
+                            _this.options.default_domain
+                        ),
+                    },
+                ],
+            });
+        },
+        get_domain: function (selected_ids) {
+            let group_domain = [];
+            const search_data = this.viewController.renderer.state;
+            let domain = search_data.domain;
+            if (this.$(".o_list_record_selector input").prop("checked")) {
+                if (search_data.groupedBy.length) {
+                    group_domain = _.filter(search_data.data, (x) => {
+                        return x.res_ids.length;
+                    }).map((x) => {
+                        return x.domain;
+                    });
+                    group_domain = _.flatten(group_domain, true);
+                    // Compute domain difference
+                    _.each(domain, (d) => {
+                        group_domain = _.without(
+                            group_domain,
+                            _.filter(group_domain, (x) => {
+                                return _.isEqual(x, d);
+                            })[0]
+                        );
+                    });
+                    // Strip operators to leave just the group domains
+                    group_domain = _.without(group_domain, "&");
+                    // Add OR operators if there is more than one group
+                    group_domain = _.times(
+                        group_domain.length - 1,
+                        _.constant("|")
+                    ).concat(group_domain);
+                }
+            } else {
+                const ids = selected_ids.map((x) => {
+                    return x.id;
+                });
+                domain = domain.concat([["id", "in", ids]]);
+            }
+            return domain.concat(group_domain);
+        },
```

Comment:
You could do something like
```suggestion
    get_domain: function (selected_ids) {
        const search_data = this.viewController.renderer.state;
        const domain = search_data.domain;
        if (
            this.$(".o_list_record_selector input").prop("checked")
            && search_data.groupedBy.length
        ) {
            const group_domain = search_data.data
                .filter((x) => x.res_ids.length)
                .map((x) => x.domain)
                .flat()
                .filter((x) => !domain.some((d) => _.isEqual(x, d)))
                .filter((x) => x != "&");
            if (group_domain.length > 1) {
                group_domain.unshift(Array(group_domain.length - 1).fill("|"));
            }
            return domain.concat(group_domain);
        }
        return domain.concat([["id", "in", selected_ids.map((x) => x.id)]]);
    },
```
this.

Reply:
@Du-ma I tested your suggestion but domain is more complex, please see my video

https://user-images.githubusercontent.com/7775116/151639804-493765ed-fcae-4dd8-a650-2bdb794f6aa1.mp4

Reply:
Sure, ill try refactoring it again when I have a free moment :grin:
Have a feeling that this is just way too over complicated than it needs to be.